text/0249-v2-experiments.md

```diff
@@ -374,6 +380,257 @@ Disadvantages:
+  with a ~~strikethrough~~, they will still be listed by autocomplete.
+- Will force some pretty long and ugly names on our APIs. Many experimental APIs
+  will result in a less aesthetic user code.
+- Graduating a module will require a lot of code changes from our users to
+  remove the prefix/suffix.
+- A lot of deprecated code in aws-cdk-lib, possibly blowing up `aws-cdk-lib`.
+  This might not be a real concern as we can reuse a lot the code between
+  different version of an API.
+
+**How can we encourage users to upgrade to a newer version?**
+
+When a new minor version introduces a breaking change to an experimental API,
+users have a clear motivation to upgrade to the new version of the API - they
+must do so in order to upgrade to a newer version of the AWS CDK. If the API is
+only deprecated, users have no motivation to upgrade to a newer version of the
+API, even worse, they might not be aware that a newer version exists. One way to
+encourage users to upgrade to a newer version of an API, is to supply tools that
+will inform users. For example, we can add a capability to the `cdk doctor`
+command that will notify users that their CDK application is using deprecated
+experimental APIs.
+
+Executing `cdk doctor` will print:
+
+```bash
+neta@dev/my-cdk-application$ cdk doctor
+
+Newer version of experimental APIs your application is using are available. You should consider upgrading.
+To see which experimental APIs can be upgraded, execute `cdk --show-deprecated-api-usage`
+```
+
+**If there are no breaking changes, are the APIs really experimental?**
+
+Given that experimental APIs are safe to use, and no breaking changes will be
+introduced, we might consider not referring to them as experimental, and simply
+add versioning in the API name. The first version of `grantWrite` will be named
+`grantWriteV1`, the second version will be named `grantWriteV2` and so on. While
+the API will not break, we should still add `experimental` for the following
+reasons:
+
+1. **Real-estate**: `grantWrite` is a much better name than `grantWriteV3` for
+   the final API.
+2. **Encourage feedback**: Declaring an API as experimental encourage the
+   community to supply feedback.
+3. **Setting the right expectations**: When an API is marked as experimental
+   users are aware that this is not the final version of the API, and its
+   deprecation is expected.
+
+### Open questions
+
+If we decide to implement option 6, we will need to formalize the experimental
+naming strategy.
+
+- Should experimental modules be named with `experimental`? e.g `aws-synthetics`
+  be named `aws-synthetics-pre-v1`, or is it sufficient that its types be named
+  as experimental, e.g `CanaryPreV1`.
+- If a type that contains other types is experimental, is it sufficient that the
+  containing type is named as `experimental` or should its properties/methods
+  should be named as experimental as well. e.g if the `Canary` construct is
```

Comment:
I think that if a type is named `XxxPre1` then members don't need to also have the postfix. This will also reduce the amount of changes required when an API is finalized.